### ISSUE TITLE
fs::write fails on long file name in deep directory path on windows

### ISSUE THREAD

**Issue by dsherret** [rust-lang/rust#76586]
Run the following code and ensure the application has access to the folder.

```rust
let dir_path = std::path::PathBuf::from(r#"C:\Users\testing.user\.bvm\bins\nodejs-node-8.12.0\node-v8.12.0-win-x64\node_modules\npm\test\npm_cache\_cacache\content-v2\sha512\32\c8\"#);
std::fs::create_dir_all(&dir_path).unwrap(); // ok
std::fs::write(&dir_path.join(r#"6992ba4671408a292bb3f2fae4cd10416dcedb6a214c169054af6bcc792b6ea56f7245333357cc59e4f84660380604b5ff338258ad46973218d864799b5e"#), "test").unwrap(); // error
```

Expected behaviour: It should create the file as I can create the file in file explorer and using other utilities like 7zip.

Actual behaviour: It errors with `Os { code: 3, kind: NotFound, message: "The system cannot find the path specified." }`

### Meta

`rustc +nightly --version --verbose`:

```
rustc 1.48.0-nightly (e2be5f568 2020-09-09)
binary: rustc
commit-hash: e2be5f568d1f60365b825530f5b5cb722460591b
commit-date: 2020-09-09
host: x86_64-pc-windows-msvc
release: 1.48.0-nightly
LLVM version: 11.0
```

### Other Info

The following equivalent application works fine in C# on the same system:

```csharp
var dirName = @"C:\Users\testing.user\.bvm\bins\nodejs-node-8.12.0\node-v8.12.0-win-x64\node_modules\npm\test\npm_cache\_cacache\content-v2\sha512\32\c8\";
Directory.CreateDirectory(dirName);
File.WriteAllText(Path.Join(dirName, "6992ba4671408a292bb3f2fae4cd10416dcedb6a214c169054af6bcc792b6ea56f7245333357cc59e4f84660380604b5ff338258ad46973218d864799b5e"), "test");
```

Edit: It occurs after the path length exceeds 260 characters.

**Comment by the8472** [comment]
* Have you enabled [long path mode](https://superuser.com/questions/1119883/windows-10-enable-ntfs-long-paths-policy-option-missinghttps://superuser.com/questions/1119883/windows-10-enable-ntfs-long-paths-policy-option-missing) in windows?
* have you tried using the [verbatim path prefix](https://docs.microsoft.com/en-us/windows/win32/fileio/naming-a-file#maximum-path-length-limitation), see also [std::Path::Prefix](https://doc.rust-lang.org/std/path/enum.Prefix.html)?

**Comment by dsherret** [comment]
@the8472 long paths are enabled.

It works with the `\\?\` prefix though. The equivalent C# application above did not need the `\\?\` prefix. I'm guessing they might add it under the hood?

**Comment by the8472** [comment]
Perhaps, or C# has an application manifest that sets the [`longPathAware` property](https://docs.microsoft.com/en-us/windows/win32/fileio/naming-a-file#enable-long-paths-in-windows-10-version-1607-and-later). I don't know if that's added when you build a rust binary.

**Comment by mati865** [comment]
> I don't know if that's added when you build a rust binary.

It's not.

I think C# might be using UNC path so this works. Can you try with C++?

**Comment by dsherret** [comment]
It seems C++ fails silently...

```c++
#include <iostream>
#include <fstream>

using namespace std;

int main()
{
    ofstream myFile("C:\\Users\\testing.user\\.bvm\\bins\\nodejs-node-8.12.0\\node-v8.12.0-win-x64\\node_modules\\npm\\test\\npm_cache\\_cacache\\content-v2\\sha512\\32\\c8\\6992ba4671408a292bb3f2fae4cd10416dcedb6a214c169054af6bcc792b6ea56f7245333357cc59e4f84660380604b5ff338258ad46973218d864799b5e");
    myFile << "test";
    myFile.close();
    std::cout << "Wrote\n"; // this outputs, but the file is not created

    return 0;
}
```

By the way, I'm starting to think this is just something that should be handled manually be the developer when working with paths. I'm remembering now that `std::fs::canonicalize` returns the UNC path so I could just use that.

**Comment by ChrisDenton** [comment]
> Perhaps, or C# has an application manifest that sets the [`longPathAware` property](https://docs.microsoft.com/en-us/windows/win32/fileio/naming-a-file#enable-long-paths-in-windows-10-version-1607-and-later). I don't know if that's added when you build a rust binary.

Unfortunately Rust does not support application manifests at all therefore you'd have to use a custom linker option. Although even this only helps if you always control the environment your application will run in.

**Comment by Lonami** [comment]
> I think this issue can be closed.

Maybe the documentation could be updated to add a note on Windows that, depending on the application, it might be necessary to canonalize very long paths for it to work.

**Comment by the8472** [comment]
The question would be where to place it, the issue is likely not unique to `fs::write`

**Comment by Lonami** [comment]
`Path`? "Note that file operations using may fail on Windows if the path is too long, and it should be [`canonicalize`](https://doc.rust-lang.org/stable/std/path/struct.Path.html#method.canonicalize)d first".

**Comment by mati865** [comment]
`canonicalize` should be used with care though.
Not only UNC paths breaks various external tools but one the most surprising limitation is not being able to use `..` (it works with standard paths).

**Comment by ChrisDenton** [comment]
Canonicalize won't work in any case. It requires opening the path before you can canonicalize the path. Opening the path was the issue in the first place so this is a chicken and egg situation.

**Comment by Lonami** [comment]
I was not aware of these caveats, I was basing myself on this:

> I'm remembering now that `std::fs::canonicalize` returns the UNC path so I could just use that

**Comment by retep998** [comment]
`canonicalize` is generally the wrong thing to use almost all the time. It's just convenient to reach for as it is the closest approximation `std` provides to what you actually need, and yet it is almost always the wrong tool.

There are several ways to improve this situation (and are by no means mutually exclusive):

* First class support in Rust for Windows manifests, and by default provide long path awareness in the manifest.
* Provide a function that turns any path into an absolute path and make people use it instead of `canonicalize`.
* Make `PathBuf::push`/`Path::join` actually be smart when relative paths are pushed onto a `\\?\` path. `Path::new(r"\\?\C:\foo").join("../bar")` should result in `\\?\C:\bar` but it currently results in `\\?\C:\foo\../bar` which is *extremely* wrong.
* Have every `std` function that takes a path automatically convert any paths over `MAX_PATH` to `\\?\` paths.
* Provide better documentation and more warnings about path length limitations.

**Comment by vallentin** [comment]
Just encountered the same issue.

I found a partial "workaround", but it only applies to reading and not writing.

As an example, let's say we have a file named `aaa...aaa.txt` (300 `a`s). Then inspecting the directory using `dir /X` in the CMD reveals the "[short names]" of the files and directories. The short name of that file would be e.g. `AAAAAA~1.TXT`. Then doing `fs::read_to_string("AAAAAA~1.TXT")` works as expected.

[short names]: https://docs.microsoft.com/en-us/windows/win32/fileio/naming-a-file#short-vs-long-names

-----

Oddly enough, using the verbatim path prefix resulted in another error for me:

```
Os {
    code: 123,
    kind: Other,
    message: "The filename, directory name, or volume label syntax is incorrect.",
}
```

**Comment by ChrisDenton** [comment]
This specific issue should hopefully be fixed in nightly by #89174, which is currently scheduled for the 1.58.0 release. This implements @retep998 fourth bullet point (automatically handles long paths). #89270 implemented the third (`path.push` works on verbatim paths). The other points are not yet done.

**Comment by ChrisDenton** [comment]
Closing as fixed. The original issues has been addressed, there is also now an (unstable) [`std::fs::absolute`](https://doc.rust-lang.org/std/path/fn.absolute.html) and the only outstanding item would be adding application manifests to binaries, which should be its own separate issue.
